docs/appconfiguration_generator_refact_proposal.md

```diff
@@ -0,0 +1,143 @@
+        monitoringAnnotations["prometheus.io/scrape"] = "true"
+        monitoringAnnotations["prometheus.io/scheme"] = g.monitoring.Scheme
+        monitoringAnnotations["prometheus.io/path"] = g.monitoring.Path
+        monitoringAnnotations["prometheus.io/port"] = g.monitoring.Port
+    }
+}
+```
+**Example 2:**
+
+The `DataBaseGenerator` needs to inject env in the corresponding resources of the WorkloadGenerator. The current implementation requires the `DataBaseGenerator` to be executed before the `WorkloadGenerator`.
+```go
+func (g *databaseGenerator) injectSecret(secret *v1.Secret) error {
+    // Inject the database information into the containers of service workload.
+    if g.workload.Service != nil {
+        for _, v := range g.workload.Service.Containers {
+            v.Env[dbHostAddressEnv] = "secret://" + secret.Name + "/hostAddress"
+            v.Env[dbUsernameEnv] = "secret://" + secret.Name + "/username"
+            v.Env[dbPasswordEnv] = "secret://" + secret.Name + "/password"
+        }
+    }
+    
+    // Inject the database information into the containers of job workload.
+    if g.workload.Job != nil {
+        for _, v := range g.workload.Job.Containers {
+            v.Env[dbHostAddressEnv] = "secret://" + secret.Name + "/hostAddress"
+            v.Env[dbUsernameEnv] = "secret://" + secret.Name + "/username"
+            v.Env[dbPasswordEnv] = "secret://" + secret.Name + "/password"
+        }
+    }
+    
+    return nil
+}
+```
+## Goal
+1. Refactor the code in the Generator section to improve its extensibility, making it easier and more maintainable to add GeneratorFunc.
+
+## Non-Goal
+1.
+## Proposal
+
+We introduce a new `XXXPatcherFunc`. The current `XXXGeneratorFunc` is used only to generate new (Kubernetes) resources, while the `XXXPatcherFunc` is used only for additional patching of resources.
+```go
+func (g *appConfigurationGenerator) Generate(spec *models.Spec) error {
+	// Generator logic only generates new resources
+	gfs := []appconfiguration.NewGeneratorFunc{
+		NewNamespaceGeneratorFunc(g.project.Name),
+		accessories.NewDatabaseGeneratorFunc(g.project, g.stack, g.appName, g.app.Workload, g.app.Database),
+		...
+	}
+	if err := appconfiguration.CallGenerators(spec, gfs...); err != nil {
+		return err
+	}
+	
+	// Patcher logic patches generated resources
+	pfs := []appconfiguration.NewPatcherFunc{
+		trait.NewPatcherFunc(g.project, g.stack, g.appName, g.app),
+	}
+	if err := appconfiguration.CallPatchers(spec.Resources.KubernetesKinds(), pfs...); err != nil {
+		return err
+	}
+	return nil
+}
+```
+
+
+### Render Order
+Perform all `GeneratorFunc` tasks first, followed by all `PatcherFunc` tasks.
+### Patcher Logic
+First, classify the resources in the spec by resource type (if it is a Kubernetes type), and perform patching on the specified resource types. For example, the opsRulePatcher needs to patch the MaxUnavailable field for resources of the deployment type.
```

Comment:
why should we separate K8s resources and TF resources? Can we process them in a unified way?

Reply:
We don't have to separate K8S resources and TF resources, I'll fix this.